llvm/test/CodeGen/X86/tailcall-caller-nocsr.ll

```diff
@@ -0,0 +1,34 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc < %s -mtriple=x86_64-linux-gnu -mattr=-sse,-avx | FileCheck %s
+
+@.str = private unnamed_addr constant [6 x i8] c"%d %d\00", align 1
+
+define void @caller(i32 %0, i32 %1) #0 {
```

Comment:
`rdi/rsi` should also be preserved according to GCC? https://godbolt.org/z/W6dzY98so

Reply:
The [attribute semantic](https://clang.llvm.org/docs/AttributeReference.html#no-caller-saved-registers) is slightly different for Clang:
> All registers are callee-saved except for registers used for passing parameters to the function or returning parameters from the function. The compiler saves and restores any modified registers that were not used for passing or returning arguments to the function.

Reply:
Okay, but it also mentions:

> Functions specified with the ‘no_caller_saved_registers’ attribute should only call other functions with the ‘no_caller_saved_registers’ attribute, or should be compiled with the ‘-mgeneral-regs-only’ flag to avoid saving unused non-GPR registers.

So is it a rational scenario? Or we should consider more like https://godbolt.org/z/Eqnh5W1en

Reply:
From what I can tell, this seems like something that ought to be done, it doesn't seem to be mandatory (from the user standpoint). Also, it looks orthogonal to the current issue, i.e., we shouldn't tail call regardless of the presence of the attribute in the callee or compiled with `-mgeneral-regs-only` (note that, unlike GCC, Clang is fine without -mgeneral-regs-only).

Reply:
I think Clang should simply report error if no `-mgeneral-regs-only` like GCC. But I agree this would be an orthogonal issue.